File: android/app/src/main/java/com/soyxan/sidesubs/PlaybackClock.java
package com.soyxan.sidesubs;

import android.os.SystemClock;
import java.util.HashMap;
import java.util.Map;

final class PlaybackClock {
    private static final double RAW_CHANGE_EPSILON = 0.20;
    private static final double SEEK_THRESHOLD = 2.5;

    private static final class State {
        double position;
        double rawPosition;
        long timeMs;
        String playbackState;
        Double pendingBackwardRaw;
    }

    private final Map<String, State> states = new HashMap<>();

    synchronized double smooth(String key, double rawPosition, String playbackState) {
        long now = SystemClock.elapsedRealtime();
        String normalized = playbackState == null ? "unknown" : playbackState.toLowerCase();
        State previous = states.get(key);

        if (!"playing".equals(normalized)) {
            State next = new State();
            next.position = rawPosition;
            next.rawPosition = rawPosition;
            next.timeMs = now;
            next.playbackState = normalized;
            states.put(key, next);
            return rawPosition;
        }

        double position;
        Double pending = null;

        if (previous == null || !"playing".equals(previous.playbackState)) {
            position = rawPosition;
        } else {
            double elapsed = Math.max(0, now - previous.timeMs) / 1000.0;
            double predicted = previous.position + elapsed;
            double rawChange = rawPosition - previous.rawPosition;
            pending = previous.pendingBackwardRaw;
            boolean confirmedBackward = false;

            if (pending != null) {
                double lowerProgress = rawPosition - pending;
                boolean stillBelowClock = predicted - rawPosition > SEEK_THRESHOLD;
                if (lowerProgress >= 0.5 && stillBelowClock) {
                    position = rawPosition;
                    confirmedBackward = true;
                    pending = null;
                } else {
                    position = predicted;
                    if (rawPosition >= predicted - SEEK_THRESHOLD) pending = null;
                }
            } else {
                position = predicted;
            }

            if (!confirmedBackward) {
                if (Math.abs(rawChange) <= RAW_CHANGE_EPSILON) {
                    position = predicted;
                } else if (rawChange < -SEEK_THRESHOLD) {
                    position = predicted;
                    pending = rawPosition;
                } else if (rawChange < 0) {
                    position = predicted;
                } else if (rawChange > elapsed + SEEK_THRESHOLD) {
                    position = rawPosition;
                } else {
                    position = Math.max(predicted, rawPosition);
                }
            }
        }

        State next = new State();
        next.position = position;
        next.rawPosition = rawPosition;
        next.timeMs = now;
        next.playbackState = normalized;
        next.pendingBackwardRaw = pending;
        states.put(key, next);
        return position;
    }

    synchronized void clear() {
        states.clear();
    }
}

File: android/app/src/main/java/com/soyxan/sidesubs/PlaybackClock.kt
package com.soyxan.sidesubs

import android.os.SystemClock
import kotlin.math.abs
import kotlin.math.max

class PlaybackClock {
    private data class State(
        val position: Double,
        val rawPosition: Double,
        val timeMs: Long,
        val playbackState: String,
        val pendingBackwardRaw: Double? = null,
    )

    private val states = mutableMapOf<String, State>()

    @Synchronized
    fun smooth(key: String, rawPosition: Double, playbackState: String?): Double {
        val now = SystemClock.elapsedRealtime()
        val normalized = playbackState?.lowercase() ?: "unknown"
        val previous = states[key]

        if (normalized != "playing") {
            states[key] = State(rawPosition, rawPosition, now, normalized)
            return rawPosition
        }

        var pending: Double? = null
        val position = if (previous == null || previous.playbackState != "playing") {
            rawPosition
        } else {
            val elapsed = max(0L, now - previous.timeMs) / 1000.0
            val predicted = previous.position + elapsed
            val rawChange = rawPosition - previous.rawPosition
            pending = previous.pendingBackwardRaw

            var confirmedBackward = false
            var resolved = predicted

            pending?.let { lower ->
                val lowerProgress = rawPosition - lower
                val stillBelowClock = predicted - rawPosition > SEEK_THRESHOLD
                if (lowerProgress >= 0.5 && stillBelowClock) {
                    resolved = rawPosition
                    confirmedBackward = true
                    pending = null
                } else if (rawPosition >= predicted - SEEK_THRESHOLD) {
                    pending = null
                }
            }

            if (!confirmedBackward) {
                resolved = when {
                    abs(rawChange) <= RAW_CHANGE_EPSILON -> predicted
                    rawChange < -SEEK_THRESHOLD -> {
                        pending = rawPosition
                        predicted
                    }
                    rawChange < 0 -> predicted
                    rawChange > elapsed + SEEK_THRESHOLD -> rawPosition
                    else -> max(predicted, rawPosition)
                }
            }
            resolved
        }

        states[key] = State(position, rawPosition, now, normalized, pending)
        return position
    }

    @Synchronized
    fun clear() = states.clear()

    private companion object {
        const val RAW_CHANGE_EPSILON = 0.20
        const val SEEK_THRESHOLD = 2.5
    }
}
